spark/dl/src/main/scala/com/intel/analytics/bigdl/optim/PredictionService.scala

```diff
@@ -0,0 +1,274 @@
+/*
+ * Copyright 2016 The BigDL Authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.intel.analytics.bigdl.optim
+
+import java.util.concurrent.LinkedBlockingQueue
+
+import com.intel.analytics.bigdl.Module
+import com.intel.analytics.bigdl.nn.abstractnn.Activity
+import com.intel.analytics.bigdl.serialization.Bigdl.AttrValue.ArrayValue
+import com.intel.analytics.bigdl.serialization.Bigdl.{AttrValue, BigDLTensor, DataType, TensorStorage}
+import com.intel.analytics.bigdl.tensor.Tensor
+import com.intel.analytics.bigdl.tensor.TensorNumericMath.TensorNumeric
+import com.intel.analytics.bigdl.tensor.TensorNumericMath.TensorNumeric.{NumericBoolean, NumericChar, NumericDouble, NumericFloat, NumericInt, NumericLong, NumericString}
+import com.intel.analytics.bigdl.utils.serializer.converters.DataConverter
+import com.intel.analytics.bigdl.utils.serializer.{DeserializeContext, ModuleSerializer, ProtoStorageType, SerializeContext}
+import com.intel.analytics.bigdl.utils._
+
+import scala.collection.JavaConverters._
+import scala.collection.mutable
+import scala.reflect.ClassTag
+import scala.reflect.runtime.universe.Type
+import scala.util.{Failure, Success, Try}
+
+/**
+ * <h6>Thread-safe Prediction Service for Concurrent Calls</h6>
+ * In this service, concurrency is kept not greater than [[numThreads]] by a `BlockingQueue`,
+ * which contains available model instances.
+ * <br><br>
+ * [[numThreads]] model instances sharing weights/bias
+ * will be put into the `BlockingQueue` during initialization.
+ * <br><br>
+ * When predict method called, service will try to take an instance from BlockingQueue`,
+ * which means if all instances are on serving, the predicting request will be blocked until
+ * some instances are released.
+ * <br><br>
+ * If exceptions caught during predict,
+ * a scalar Tensor[String] will be returned with thrown message.
+ *
+ * @param model BigDL model used to do predictions
+ * @param numThreads max concurrency
+ */
+class PredictionService[T: ClassTag] private[optim](
+  model: Module[T],
+  numThreads: Int
+)(implicit ev: TensorNumeric[T]) {
+
+  initThreads()
+
+  private def initThreads(): Unit = {
+    Engine.setEngineType(LocalEngine)
```

Comment:
You can init an localMode Engine
```
def init: Unit = this.synchronized {
    if (localMode) {
      logger.info("Detect bigdl.localMode is set. Run workload without spark")
      // The physical core number should have been initialized
      // by java property -Dbigdl.coreNumber=xx
      setNodeAndCore(1, getCoreNumberFromProperty)
    } else {
      logger.info("Auto detect executor number and executor cores number")
      val (nExecutor, executorCores) = sparkExecutorAndCore().get
      logger.info(s"Executor number is $nExecutor and executor cores number is $executorCores")
      setNodeAndCore(nExecutor, executorCores)
      checkSparkContext
    }
  }
```

Reply:
If localMode, the modelPoolSize always equals to 1 when engineType == MKlBlas by defaults.

    private def initThreadPool(core : Int) : Unit = {
      val defaultPoolSize: Int = System.getProperty("bigdl.utils.Engine.defaultPoolSize",
        (core * 50).toString).toInt
      if(_default == null || _default.getPoolSize != defaultPoolSize) {
        _default = new ThreadPool(defaultPoolSize)
      }
  
      val modelPoolSize: Int = if (engineType == MklBlas) {
        1
      } else {
        core
      }
  
      if(_model == null || _model.getPoolSize != modelPoolSize) {
        _model = new ThreadPool(modelPoolSize)
        _model.setMKLThread(MKL.getMklNumThreads)
      }
    }

Reply:
If modelPoolSize=1, will just run the model in current thread.